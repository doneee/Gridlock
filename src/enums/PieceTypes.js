export default {
  Free: 0,
  Used: 1,
  Block: 2
};

export const PieceClasses = {
  0: 'piece-free',
  1: 'piece-used',
  2: 'piece-block'
};