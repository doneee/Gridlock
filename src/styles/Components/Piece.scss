.piece {
  display: inline-block;
  box-sizing: border-box;

  height: 50px;
  width: 50px;

  border: solid 1px #000;
  background: #fff;

  &.piece-used {
    background: #ccc;
  }

  &.piece-block {
    background: #777;
  }

  &.piece-active {
    position: relative;
    z-index: 2;
  }

  .move {
    position: absolute;
    box-shadow: inset 0 0 20px 6px lightblue;

    box-sizing: border-box;
    border: solid 1px #000;
    margin: -1px 0 0 -1px;

    &.up {
      top: calc(-100% - 2px);
    }
    &.down {
      top: calc(100% + 2px);
    }
    &.left {
      left: calc(-100% - 2px);
    }
    &.right {
      left: calc(100% + 2px);
    }
  }
}

.playfield.unstarted .piece.piece-active {
  box-shadow: inset 0 0 10px 3px lightblue;
}

.playfield:not(.unstarted) .piece.piece-used.piece-active {
  background: #333;
}

.playfield.loss .piece.piece-free {
  background: #f00;
}