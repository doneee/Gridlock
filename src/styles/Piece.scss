.piece {
  display: inline-block;
  box-sizing: border-box;

  height: 50px;
  width: 50px;

  border: solid 1px #000;

  &.piece-used {
    background: #ccc;
  }

  &.piece-block {
    background: #777;
  }

  &.piece-active {
    position: relative;
    z-index: 2;
  }

  .move {
    position: absolute;
    height: 48px;
    width: 48px;
    box-shadow: inset 0 0 20px 6px lightblue;

    &.up {
      top: -50px;
    }
    &.down {
      top: 50px;
    }
    &.left {
      left: -50px;
    }
    &.right {
      left: 50px;
    }
  }
}

.playfield.unstarted .piece.piece-active {
  box-shadow: inset 0 0 10px 3px lightblue;
}

.playfield:not(.unstarted) .piece.piece-used.piece-active {
  background: #333;
}

.playfield.loss .piece.piece-free {
  background: #f00;
}