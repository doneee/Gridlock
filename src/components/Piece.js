import React, {PropTypes, Component} from 'react';

import PieceTypes, {PieceClasses} from '../enums/PieceTypes';
import GAME_STATE from '../enums/GameState';

import '../styles/components/Piece.scss';

class Piece extends Component {

  constructor(props) {
    super(props);

    this.handleTouchTap = this.handleTouchTap.bind(this);
    this.handleDirectionTap = this.handleDirectionTap.bind(this);
  }

  handleTouchTap (e) {
    this.props.pieceTapped(this.props.position, 1);
  }

  handleDirectionTap (direction, e) {
    this.props.directionTapped(direction);
  }

  render () {
    let {gameState, pieceState, pieceActive, pieceDimensions} = this.props;

    let directions = this.props.getValidDirections(this.props.position);
    let directionButtons = [];

    if ((gameState === GAME_STATE.STARTED) && pieceActive) {
      directionButtons = Object.keys(directions)
        .map((direction) => directions[direction] ? 
          <div key={direction} 
            className={`move ${direction}`}
            style={{height: pieceDimensions[1], width: pieceDimensions[0]}}
            onMouseDown={() => this.handleDirectionTap(direction)} />
        : '');      
    }

    return (
      <div 
        className={`piece ${PieceClasses[pieceState]} ${pieceActive ? 'piece-active' : ''}`}
        style={{height: pieceDimensions[1], width: pieceDimensions[0]}}
        onMouseDown={this.handleTouchTap}>
          {directionButtons}
      </div>
    );
  }
}

Piece.propTypes = {
  position: PropTypes.arrayOf(PropTypes.number),
  pieceState: PropTypes.oneOf(Object.values(PieceTypes)),
  pieceActive: PropTypes.bool,
  pieceTapped: PropTypes.func,
  directionTapped: PropTypes.func,
  rows: PropTypes.number,
  columns: PropTypes.number,
  gameState: PropTypes.oneOf(Object.values(GAME_STATE)),
  getValidDirections: PropTypes.func
};

Piece.defaultProps = {
  position: [0, 0],
  pieceState: 0,
  pieceActive: false,
  pieceDimensions: [0, 0],
  pieceTapped: () => {},
  directionTapped: () => {},
  rows: 0,
  columns: 0,
  gameState: GAME_STATE.UNSTARTED,
  getValidDirections: () => ({ up: true, down: true, left: true, right: true })
}

export default Piece;