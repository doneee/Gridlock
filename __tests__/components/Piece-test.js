// Link.react-test.js
import React from 'react';
import renderer from 'react-test-renderer';

import Piece from '../../src/components/Piece';
import PieceTypes from '../../src/enums/PieceTypes';
import GAME_STATE from '../../src/enums/GameState';

describe('<Piece />', () => {
  it('renders with default props', () => {
    const component = renderer.create(
      <Piece />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with active state', () => {
    const component = renderer.create(
      <Piece pieceActive={true} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('renders with used state', () => {
    const component = renderer.create(
      <Piece pieceState={PieceTypes.Used} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders direction buttons when active and game is started', () => {
    const component = renderer.create(
      <Piece pieceActive={true} gameState={GAME_STATE.STARTED} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
