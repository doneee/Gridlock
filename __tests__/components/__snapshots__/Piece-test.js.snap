exports[`<Piece /> renders direction buttons when active and game is started 1`] = `
<div
  className="piece piece-free piece-active"
  onMouseDown={[Function]}
  style={
    Object {
      "height": 0,
      "width": 0,
    }
  }>
  <div
    className="move up"
    onMouseDown={[Function]}
    style={
      Object {
        "height": 0,
        "width": 0,
      }
    } />
  <div
    className="move down"
    onMouseDown={[Function]}
    style={
      Object {
        "height": 0,
        "width": 0,
      }
    } />
  <div
    className="move left"
    onMouseDown={[Function]}
    style={
      Object {
        "height": 0,
        "width": 0,
      }
    } />
  <div
    className="move right"
    onMouseDown={[Function]}
    style={
      Object {
        "height": 0,
        "width": 0,
      }
    } />
</div>
`;

exports[`<Piece /> renders with active state 1`] = `
<div
  className="piece piece-free piece-active"
  onMouseDown={[Function]}
  style={
    Object {
      "height": 0,
      "width": 0,
    }
  } />
`;

exports[`<Piece /> renders with default props 1`] = `
<div
  className="piece piece-free "
  onMouseDown={[Function]}
  style={
    Object {
      "height": 0,
      "width": 0,
    }
  } />
`;

exports[`<Piece /> renders with used state 1`] = `
<div
  className="piece piece-used "
  onMouseDown={[Function]}
  style={
    Object {
      "height": 0,
      "width": 0,
    }
  } />
`;
